flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImplTest.java

```diff
@@ -0,0 +1,357 @@
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.finish();
+		}
+
+		assertTrue(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+		jobAwareEnvironment.releaseAllFinishedPartitionsForJobAndMarkJobInactive(jobId);
+		assertThat(wrappedEnvironment.releasedPartitions, contains(externallyManagedWriter.resultPartitionID));
+	}
+
+	@Test
+	public void testReleasePartitions() throws IOException {
+		jobAwareEnvironment.markJobActive(jobId);
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.finish();
+		}
+
+		assertTrue(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+		jobAwareEnvironment.releaseFinishedPartitions(jobId, Collections.singleton(externallyManagedWriter.resultPartitionID));
+		assertThat(wrappedEnvironment.releasedPartitions, contains(externallyManagedWriter.resultPartitionID));
+	}
+
+	@Test
+	public void testPartitionFailure() throws IOException {
+		jobAwareEnvironment.markJobActive(jobId);
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.fail(null);
+		}
+
+		assertFalse(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+		jobAwareEnvironment.releaseAllFinishedPartitionsForJobAndMarkJobInactive(jobId);
+		assertThat(wrappedEnvironment.releasedPartitions, empty());
+	}
+
+	@Test
+	public void testPartitionFinishWithoutActiveJob() throws IOException {
+		for (ResultPartitionWriter writer : writers) {
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.finish();
+		}
+
+		assertFalse(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+		assertThat(wrappedEnvironment.releasedPartitions, contains(externallyManagedWriter.resultPartitionID));
+	}
+
+	@Test
+	public void testPartitionFinishWithActiveJob() throws IOException {
+		jobAwareEnvironment.markJobActive(jobId);
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.finish();
+		}
+
+		assertTrue(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+		assertThat(wrappedEnvironment.releasedPartitions, empty());
+	}
+
+	@Test
+	public void testPartitionFinishWithPreviouslyActiveJob() throws IOException {
+		jobAwareEnvironment.markJobActive(jobId);
+
+		jobAwareEnvironment.releaseAllFinishedPartitionsForJobAndMarkJobInactive(jobId);
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.finish();
+		}
+
+		assertFalse(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+		assertThat(wrappedEnvironment.releasedPartitions, contains(externallyManagedWriter.resultPartitionID));
+	}
+
+	private static Collection<ResultPartitionWriter> createResultPartitionWriters(
```

Comment:
This utility method seems not very necessary atm because it is only used in one place.

Reply:
it is used twice though in that place.

Reply:
it seems only used once in `setup`.

Reply:
ah, I was looking at the wrong method!